conda_build/metadata.py

```diff
@@ -863,48 +878,40 @@ def toposort(output_metadata_map):
         for output_d in output_metadata_map
         if output_d.get("type", "conda").startswith("conda")
     ]
-    topodict = dict()
-    order = dict()
-    endorder = set()
+    topodict: dict[str, set[str]] = {}
+    name_lookup: dict[str, list[OutputTuple]] = {}
+    extra: list[OutputTuple] = []
 
-    for idx, (output_d, output_m) in enumerate(output_metadata_map.items()):
+    for idx, (output_d, output_m) in enumerate(output_metadata_map.values()):
         if output_d.get("type", "conda").startswith("conda"):
-            deps = output_m.get_value("requirements/run", []) + output_m.get_value(
-                "requirements/host", []
-            )
-            if not output_m.is_cross:
-                deps.extend(output_m.get_value("requirements/build", []))
+            # the same package name may be seen multiple times (variants)
             name = output_d["name"]
-            order[name] = idx
-            topodict[name] = set()
-            for dep in deps:
-                dep = dep.split(" ")[0]
-                if dep in these_packages:
-                    topodict[name].update((dep,))
+
+            # dependencies for all of the variants
+            dependencies = (
+                *output_m.get_value("requirements/run", []),
+                *output_m.get_value("requirements/host", []),
+                *(
+                    output_m.get_value("requirements/build", [])
+                    if not output_m.is_cross
+                    else []
+                ),
+            )
+            topodict.setdefault(name, set()).update(
+                dependency_name
+                for dependency in dependencies
+                if (dependency_name := MatchSpec(dependency).name) in these_packages
```

Comment:
Same question as with `check_circular_dependencies` here: Does `MatchSpec` always work at this point? If yes, then all good!

Reply:
the code was far from clear so to be safe I've reverted the `MatchSpec` change: https://github.com/conda/conda-build/pull/5345/commits/45232e1d17e4608fa7c4bd0c0f16270755662ac0

Reply:
Bummer, but thanks for checking!
Let's hope some of these code paths can be entangled/replaced some time in the future so we can safely apply such improvements again! :crossed_fingers:

Reply:
Agreed